File: csharp/DataStructures/src/dsa/sorting/Sort012.cs
namespace dsa.sorting;

public class Sort012
{
    // static void Main()
    // {
    //     int[] arr = new int[] { 2, 0, 0, 1, 0, 2, 0, 1, 0, 2, 2, 0 };

    //     Sort012 st = new();

    //     st.printArray(arr);
    //     st.threeNumberSort(arr);
    //     st.printArray(arr);
    // }
    
    public void printArray(int[] arr)
    {
        int n = arr.Length;

        for (int i = 0; i < n; i++)
        {
            Console.Write(arr[i] + " ");
        }
        Console.WriteLine();
    }

    public void threeNumberSort(int[] arr)
    {
        int i = 0;
        int j = 0;
        int k = arr.Length - 1;

        // Percorre o array
        while (i <= k)
        {
            // Se o elemento atual é 0, troca com o elemento na posição j e incrementa i e j
            if (arr[i] == 0)
            {
                swap(arr, i, j);
                j++;
                i++;
            }
            // Se o elemento atual é 1, apenas incrementa i
            else if (arr[i] == 1)
                i++;

            // Se o elemento atual é 2, troca com o elemento na posição k e decrementa k
            else if (arr[i] == 2)
            {
                swap(arr, i, k);
                k--;
            }
        }
    }

    // Método para trocar dois elementos de um array
    private static void swap(int[] arr, int first, int second)
    {
        // Armazena temporariamente o valor do primeiro elemento
        int temp = arr[first];

        // Troca os valores dos elementos
        arr[first] = arr[second];
        arr[second] = temp;
    }
}

File: go/dsa/sorting/Sort012/Sort012.go
package sort012

import "fmt"

// PrintArray imprime os elementos do array em uma única linha.
func PrintArray(arr []int) {
	for _, v := range arr { // Itera sobre cada elemento do array
		fmt.Printf("%d ", v) // Imprime o valor atual com um espaço
	}
	fmt.Println() // Quebra de linha após imprimir todos os elementos
}

// ThreeNumberSort realiza a ordenação de um array contendo apenas os números 0, 1 e 2.
// O algoritmo é conhecido como "Dutch National Flag Problem".
func ThreeNumberSort(arr []int) {
	i := 0            // Ponteiro que indica a posição atual no array
	j := 0            // Ponteiro que indica a posição do próximo 0 a ser trocado
	k := len(arr) - 1 // Ponteiro que indica a posição do próximo 2 a ser trocado

	// O loop continua enquanto o ponteiro i estiver antes ou na mesma posição que k
	for i <= k {
		// Se o valor na posição i for 0
		if arr[i] == 0 {
			// Troca o valor atual com a posição do próximo 0 (indicado por j)
			swap(arr, i, j)
			i++ // Avança o ponteiro i para o próximo elemento
			j++ // Avança o ponteiro j para a próxima posição onde o próximo 0 deve estar
		} else if arr[i] == 1 {
			// Se o valor na posição i for 1, apenas avança o ponteiro i
			i++
		} else if arr[i] == 2 {
			// Se o valor na posição i for 2
			// Troca o valor atual com a posição do próximo 2 (indicado por k)
			swap(arr, i, k)
			k-- // Diminui o ponteiro k para que o próximo 2 seja colocado corretamente
		}
	}
}

// swap troca os valores de dois elementos no array
func swap(arr []int, first int, second int) {
	temp := arr[first]    // Armazena temporariamente o valor da posição first
	arr[first] = arr[second] // Move o valor da posição second para first
	arr[second] = temp    // Coloca o valor armazenado temporariamente na posição second
}
